solc/CommandLineInterface.cpp

```diff
@@ -931,45 +932,61 @@ void CommandLineInterface::link()
 		// be just the cropped or '_'-padded library name, but this changed to
 		// the cropped hex representation of the hash of the library name.
 		// We support both ways of linking here.
-		librariesReplacements["__" + evmasm::LinkerObject::libraryPlaceholder(name) + "__"] = library.second;
-
-		string replacement = "__";
-		for (size_t i = 0; i < placeholderSize - 4; ++i)
-			replacement.push_back(i < name.size() ? name[i] : '_');
-		replacement += "__";
-		librariesReplacements[replacement] = library.second;
+		librariesReplacements[name] = library.second;
 	}
 
 	FileReader::StringMap sourceCodes = m_fileReader.sourceUnits();
+
 	for (auto& src: sourceCodes)
 	{
-		auto end = src.second.end();
-		for (auto it = src.second.begin(); it != end;)
+		// Map from how the libraries will be named inside the bytecode to their addresses.
+		map<size_t, string> linkReferences;
+		string bytecode;
+		vector<string> sourceLines;
+		boost::split(sourceLines, src.second, boost::is_any_of("\n"));
+
+		if (sourceLines.size() < 3)
+			// no bytecode found
+			continue;
+
+		size_t bytecodeIndex;
+		// Bytecode can be on line 0 or 3, depending on if file header is included
+		// However if it is a file the first line is always blank as it just contains a '\n'
+		if (boost::starts_with(sourceLines.at(1), "======="))
+			bytecodeIndex = 3;
+		else
+			bytecodeIndex = 0;
+
+		for (auto const& library: m_options.linker.libraries)
 		{
-			while (it != end && *it != '_') ++it;
-			if (it == end) break;
-			if (
-				end - it < placeholderSize ||
-				*(it + 1) != '_' ||
-				*(it + placeholderSize - 2) != '_' ||
-				*(it + placeholderSize - 1) != '_'
-			)
-				solThrow(
-					CommandLineExecutionError,
-					"Error in binary object file " + src.first + " at position " + to_string(it - src.second.begin()) + "\n" +
-					'"' + string(it, it + min(placeholderSize, static_cast<int>(end - it))) + "\" is not a valid link reference."
-				);
-
-			string foundPlaceholder(it, it + placeholderSize);
-			if (librariesReplacements.count(foundPlaceholder))
+			string const& name = library.first;
+			// Bytecode we need is on the third line there's a blank line at the start
+			bytecode = sourceLines.at(bytecodeIndex);
+			size_t libraryStartOffset = bytecode.find("__$");
+
+			for (size_t i = 0; i < placeholderSize; i++)
+				bytecode[libraryStartOffset + i] = '0';
```

Comment:
This code used to throw on invalid placeholders. Now it just looks for `__$` and assumes that anything it founds must be a valid placeholder. But it could be something else (`$__` missing) or it could be truncated and you'll go past the end of the `bytecode` array.